docs/design/secrets.md

```diff
@@ -0,0 +1,132 @@
+access secrets. Secrets should be placed where the container expects them to be.
+
+## Constraints and Assumptions
+
+* This design does not prescribe a method for storing/transmitting secrets
+* Encryption and node security are orthogonal concerns
+* It is assumed that node and master are secure and that compromising their security could also
+  compromise secrets.
+
+## Use Cases
+
+1.  As a cluster operator, I want to allow a pod to access the Kubernetes master using a custom
+    .kubeconfig, so that I can securely reach the master
+2.  As a cluster operator, I want to allow a pod to access the DockerHub using credentials from a
+    .dockercfg file, so that pods can push images to the DockerHub
+3.  As a cluster operator, I want to allow a pod to access a GitHub repository using SSH keys, so
+    that I can push and fetch to and from the repository
+4.  As a user, I want to allow pods to consume supplemental information about services such as
+    username and password which should be kept secret, so that I can share secrets about a service
+    amongst the pods in my application securely
+5.  As a user, I want to store secret configuration artifacts for my applications and consume
+    them securely in pods, so that I can keep the configuration for my applications separate from
+    the images that use them.
+
+
+## Proposed Design
+
+### Overview
+
+This design proposes mounting secrets with a new volume type. The secrets volume will be backed
+by a volume plugin that does the actual work of fetching the secret and placing it on the
+filesystem to be mounted in the container. It should be possible to mount single files as part of a
+secret. Secrets may consist of multiple files. For example, an SSH key pair. In order to remove the
+burden from the end user in specifying every file that a secret consists of, it should be possible
+to mount all files provided by a secret with a single ```VolumeMount``` entry in the container
+specification.
+
+### Secret Volume Source
+
+A new Secret type of volume source will be added to the ```VolumeSource``` struct in the API:
+
+```go
+type VolumeSource struct {
+     ... 
+     Secret *Secret `json:"secret"`
+}
+
+type Secret struct {
+     // Ref is a reference to the actual secret
+     Ref string 
+     
+     // Files is a set of descriptors of where and how the secret files
+     // should be placed (optional). If not specified, the secret itself
+     // will determine where its files should go
+	 Files []SecretFile
+}
+
+type SecretFile struct {
+     // Name of the file within the secret. For example, in the case
+     // of SSH keys, you would have a 'private' and a 'public' key
+	 Name string
+	 
+	 // Path of the secret within the container
+	 Path string
+	 
+	 // Owner id of the file owner
+	 OwnerUID int
+	 
+	 // Group id of the file group
+	 GroupUID int
+	 
+	 // File Mode
+	 Mode os.FileMode
+	 
+	 // MCSLabel for SELinux
+	 MCSLabel string
```

Comment:
It does, and I think the security related attributes (apart from Mode) of this structure should likely be removed, because we will associate a security context to the secret.